### PR TITLE
[SYCL-MLIR][NFC] Refactor CommonFieldLookup for SYCL type

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Signed-off-by: Tsang, Whitney <whitney.tsang@intel.com>

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 1]
```diff
@@ -1317,6 +1317,23 @@ ValueCategory MLIRScanner::VisitBinaryOperator(clang::BinaryOperator *BO) {
   }
 }
 
+template <typename T>
+Value MLIRScanner::SYCLCommonFieldLookup(Value V, size_t FNum,
+                                         std::vector<int64_t> shape) {
```
[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc:1322]: pass a const reference to the vector pls.

File: polygeist/tools/cgeist/Lib/clang-mlir.h [hunk 2]
```diff
@@ -281,6 +281,11 @@ class MLIRScanner : public clang::StmtVisitor<MLIRScanner, ValueCategory> {
     return UnsupportedFuncs.contains(Name);
   }
 
+  // Get the \p fnum field of Value \p V of element type T.
```
[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/clang-mlir.h:284]: Document that the type of \p V must be a memref. Also from the implementation looks like the template parameter T must be equal to the element type of the memref.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 3]
```diff
@@ -1317,6 +1317,23 @@ ValueCategory MLIRScanner::VisitBinaryOperator(clang::BinaryOperator *BO) {
...
+  auto MT = V.getType().cast<MemRefType>();
+  Type ElemTy = MT.getElementType();
+  assert(ElemTy.isa<T>() && "Expecting element type to be the templated type");
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:1325]: static_assert(std::is_same<T,ElemTy>::value, ".....");

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 4]
```diff
@@ -1404,90 +1421,36 @@ ValueCategory MLIRScanner::CommonFieldLookup(clang::QualType CT,
...
+    if (auto AT = ElemTy.dyn_cast<sycl::ArrayType>()) {
+      assert(fnum < AT.getBody().size() && "ERROR");
+      const auto ElemType = AT.getBody()[fnum].cast<MemRefType>();
+      const auto ResultType =
+          MemRefType::get(ElemType.getShape(), ElemType.getElementType(),
+                          MemRefLayoutAttrInterface(), MT.getMemorySpace());
+      Result = Builder.create<polygeist::SubIndexOp>(Loc, ResultType, val,
+                                                     getConstantIndex(fnum));
+    } else if (ElemTy.isa<sycl::AccessorType>())
+      Result = SYCLCommonFieldLookup<sycl::AccessorType>(val, fnum, shape);
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:1435]: shape -> Shape

File: polygeist/tools/cgeist/Lib/clang-mlir.cc (hunk 3 above)
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc:1325]:
```
/llvm/polygeist/tools/cgeist/Lib/clang-mlir.cc:1326:33: error: template argument for template type parameter must be a type
  static_assert(std::is_same<T, ElemTy>::value);
  ```

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 5]
```diff
@@ -1382,12 +1400,12 @@ ValueCategory MLIRScanner::CommonFieldLookup(clang::QualType CT,
           ValueCategory(commonGEP, /*isReference*/ true).getValue(Builder);
     return ValueCategory(commonGEP, /*isReference*/ true);
   }
-  auto MT = val.getType().cast<MemRefType>();
-  auto shape = std::vector<int64_t>(MT.getShape());
-  if (shape.size() > 1) {
-    shape.erase(shape.begin());
+  auto MT = Val.getType().cast<MemRefType>();
+  auto Shape = std::vector<int64_t>(MT.getShape());
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:1404]: Can we use `llvm::SmallVector` here instead?

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 6]
```diff
@@ -1396,111 +1414,57 @@ ValueCategory MLIRScanner::CommonFieldLookup(clang::QualType CT,
...
+    else if (ElemTy.isa<sycl::ItemType>())
+      Result = SYCLCommonFieldLookup<sycl::ItemType>(Val, FNum, Shape);
+    else if (ElemTy.isa<sycl::ItemBaseType>())
+      Result = SYCLCommonFieldLookup<sycl::ItemBaseType>(Val, FNum, Shape);
+    else if (ElemTy.isa<sycl::NdItemType>())
+      Result = SYCLCommonFieldLookup<sycl::NdItemType>(Val, FNum, Shape);
+    else if (ElemTy.isa<sycl::GroupType>())
+      Result = SYCLCommonFieldLookup<sycl::GroupType>(Val, FNum, Shape);
+    else
+      llvm_unreachable("not implemented");
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:1454]: NIT: Use `llvm::Typeswitch` instead.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc (hunk 5 above)
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc]: done.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 7]
```diff
@@ -1427,36 +1427,52 @@ ValueCategory MLIRScanner::CommonFieldLookup(clang::QualType CT,
...
+            })
+            .Case<sycl::ItemType>([&](auto ElemTy) {
+              return SYCLCommonFieldLookup<sycl::ItemType>(Val, FNum, Shape);
+            })
+            .Case<sycl::NdItemType>([&](auto ElemTy) {
+              return SYCLCommonFieldLookup<sycl::NdItemType>(Val, FNum, Shape);
+            })
+            .Case<sycl::NdRangeType>([&](auto ElemTy) {
+              return SYCLCommonFieldLookup<sycl::NdRangeType>(Val, FNum, Shape);
+            })
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:1471]:
```suggestion
            .Case<sycl::AccessorType, sycl::AccessorImplDeviceType,
                       sycl::AccessorSubscriptType, sycl::AtomicType,
                       sycl::GroupType, sycl::ItemBaseType,
                       sycl::ItemType, sycl::NdItemType,
                       sycl::NdRangeType>([&](auto ElemTy) {
              return SYCLCommonFieldLookup<decltype(ElemTy)>(Val, FNum, Shape);
            })
```
Sorry for being too picky, but this saves so many lines.

**Review by victor-eds — COMMENTED** [review]
LGTM after suggestion.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc (hunk 7 above)
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc:1471]: That's great, I wanted to do something like that, but didn't know `decltype`, thanks for the suggestion.

**Review by victor-eds — APPROVED** [review]

**Pull request merged**